### PR TITLE
Grid Search File Support

### PR DESCRIPTION
**Pull request by junshiguo**
https://github.com/ShifuML/shifu/issues/423
Tested reading params locally.

### CODE REVIEW

File: src/main/java/ml/shifu/shifu/util/CommonUtils.java [hunk 1]
```diff
@@ -500,8 +539,8 @@ public boolean apply(ColumnConfig input) {
             reader = ShifuFileUtils.getReader(pathHeader, sourceType);
             pigHeaderStr = reader.readLine();
             if(StringUtils.isEmpty(pigHeaderStr)) {
-                throw new RuntimeException(String.format("Cannot reade header info from the first line of file: %s",
-                        pathHeader));
```
Comment (zhangpengshan) [inline src/main/java/ml/shifu/shifu/util/CommonUtils.java]: Please use shifu-formatter.xml for code format

File: src/main/java/ml/shifu/shifu/util/CommonUtils.java [hunk 2]
```diff
@@ -248,7 +249,40 @@ public static ModelConfig loadModelConfig() throws IOException {
      *             if {@code path} is null or empty, if sourceType is null.
      */
     public static ModelConfig loadModelConfig(String path, SourceType sourceType) throws IOException {
-        return loadJSON(path, sourceType, ModelConfig.class);
+        ModelConfig modelConfig = loadJSON(path, sourceType, ModelConfig.class);
+        if(modelConfig.getTrain().getGridConfigFile() != null) {
+            // Only load file content. Grid search params parsing is done in {@link GridSearch} initialization.
+            modelConfig.getTrain().setGridConfigFileContent(loadFileContent(modelConfig.getTrain().getGridConfigFile(), sourceType));
```
Comment (zhangpengshan) [inline src/main/java/ml/shifu/shifu/util/CommonUtils.java]: One question, if configuration file is in local, how to make sure such file can also be read in HDFS mode. Like in NNMaster, it is run in cluster, can you read such configuration file from HDFS?

File: src/main/java/ml/shifu/shifu/core/dtrain/gs/GridSearch.java [hunk 3]
```diff
@@ -68,10 +80,63 @@
...
+        if(gridParams != null && gridParams.size() > 0) {
+            assert this.hasHyperParam(rawParams) == false;
+            Set<String> hyperParams = new HashSet<String>();
+            for(Map<String, Object> map: gridParams) {
+                for(Map.Entry<String, Object> gridEntry: map.entrySet()) {
+                    hyperParams.add(gridEntry.getKey());
+                }
+                for(Map.Entry<String, Object> kvEntry: this.rawParams.entrySet()) {
+                    map.put(kvEntry.getKey(), kvEntry.getValue());
+                }
```
Comment (zhangpengshan) [inline src/main/java/ml/shifu/shifu/core/dtrain/gs/GridSearch.java]: here parameter set in grid search file will be overrided by in parameters, add check to if not existing in rawParams put t to the map?

File: src/main/java/ml/shifu/shifu/core/dtrain/gs/GridSearch.java [hunk 4]
```diff
@@ -68,10 +80,63 @@
...
+                if(paramsMap != null) {
+                    gridParams.add(paramsMap);
+                }
+            }
+            return gridParams;
+        }
+        return null;
+    }
+
+    private Map<String, Object> parseParams(Map<String, MetaItem> metaWarehouse, String configLine) {
```
Comment (zhangpengshan) [inline src/main/java/ml/shifu/shifu/core/dtrain/gs/GridSearch.java]: how you can explain such format to users, document?

File: src/main/java/ml/shifu/shifu/core/dtrain/gs/GridSearch.java [hunk 5]
```diff
@@ -179,6 +244,74 @@ public boolean isGridSearchMode() {
...
+        }
+        throw new ShifuException(ShifuErrorCode.ERROR_GRID_SEARCH_FILE_CONFIG);
+    }
+
+    /**
+     * Hard Coded!
+     * 
+     * @param key
+     * @return
+     */
```
Comment (zhangpengshan) [inline src/main/java/ml/shifu/shifu/core/dtrain/gs/GridSearch.java]: @param and @return should have comments

File: src/main/java/ml/shifu/shifu/util/CommonUtils.java [hunk 6]
```diff
@@ -80,6 +80,7 @@
 import org.apache.hadoop.fs.Path;
 import org.apache.hadoop.fs.PathFilter;
 import org.apache.hadoop.fs.RemoteIterator;
+import org.apache.hadoop.yarn.webapp.hamlet.Hamlet.P;
```
Comment (zhangpengshan) [inline src/main/java/ml/shifu/shifu/util/CommonUtils.java]: P? remove it?

File: src/main/java/ml/shifu/shifu/util/CommonUtils.java [hunk 7]
```diff
@@ -248,7 +249,40 @@ public static ModelConfig loadModelConfig() throws IOException {
...
+     * @throws IOException
+     *             if any IO exception in reading file content.
+     */
+    private static List<String> loadFileContent(String path, SourceType sourceType) throws IOException {
+        checkPathAndMode(path, sourceType);
+        log.debug("loading {} with sourceType {}", path, sourceType);
+        BufferedReader reader = null;
+        try {
+            reader = ShifuFileUtils.getReader(path, sourceType);
+            List<String> content = new ArrayList<String>();
```
Comment (zhangpengshan) [inline src/main/java/ml/shifu/shifu/util/CommonUtils.java]: contents may be a better name

File: src/main/java/ml/shifu/shifu/util/CommonUtils.java (hunk 6 above)
Comment (junshiguo) [reply src/main/java/ml/shifu/shifu/util/CommonUtils.java]: Updated!

File: src/main/java/ml/shifu/shifu/util/CommonUtils.java (hunk 2 above)
Comment (junshiguo) [reply src/main/java/ml/shifu/shifu/util/CommonUtils.java]: The code reads file with sourcetype using ShifuFileUtils.getReader(path, sourceType), same as reading ModelConfig.json. I suppose that would handle file location.
I will add code to copy gridsearch config file from local to HDFS to make sure it is at the same   location as ModelConfig.json.

File: src/main/java/ml/shifu/shifu/util/CommonUtils.java (hunk 7 above)
Comment (junshiguo) [reply src/main/java/ml/shifu/shifu/util/CommonUtils.java]: updated!

File: src/main/java/ml/shifu/shifu/util/CommonUtils.java (hunk 1 above)
Comment (junshiguo) [reply src/main/java/ml/shifu/shifu/util/CommonUtils.java]: I am already using shifu-formatter.xml to format code. Eclipse gives me this...

File: src/main/java/ml/shifu/shifu/core/dtrain/gs/GridSearch.java (hunk 3 above)
Comment (junshiguo) [reply src/main/java/ml/shifu/shifu/core/dtrain/gs/GridSearch.java]: updated!

File: src/main/java/ml/shifu/shifu/core/dtrain/gs/GridSearch.java (hunk 4 above)
Comment (junshiguo) [reply src/main/java/ml/shifu/shifu/core/dtrain/gs/GridSearch.java]: updated the log to add hints

File: src/main/java/ml/shifu/shifu/core/dtrain/gs/GridSearch.java (hunk 5 above)
Comment (junshiguo) [reply src/main/java/ml/shifu/shifu/core/dtrain/gs/GridSearch.java]: added

File: src/main/java/ml/shifu/shifu/util/CommonUtils.java (hunk 2 above)
Comment (junshiguo) [reply src/main/java/ml/shifu/shifu/util/CommonUtils.java]: Please check latest commits. gridsearch config file will be at the same folder with ModelConfig.json if in HDFS.

File: src/main/java/ml/shifu/shifu/util/CommonUtils.java [hunk 8]
```diff
@@ -138,6 +138,14 @@ public static boolean copyConfFromLocalToHDFS(ModelConfig modelConfig, PathFinde
         Path dstModelConfig = new Path(pathFinder.getModelSetPath(SourceType.HDFS));
         hdfs.copyFromLocalFile(srcModelConfig, dstModelConfig);
 
+        // Copy GridSearch config file if exists
+        String gridConfigFile = modelConfig.getTrain().getGridConfigFile();
+        if(gridConfigFile != null && !gridConfigFile.trim().equals("")) {
```
Comment (zhangpengshan) [inline src/main/java/ml/shifu/shifu/util/CommonUtils.java]: Only check non empty while in fact better to have file-existing check

File: src/main/java/ml/shifu/shifu/util/CommonUtils.java [hunk 9]
```diff
@@ -248,7 +256,46 @@ public static ModelConfig loadModelConfig() throws IOException {
      *             if {@code path} is null or empty, if sourceType is null.
      */
     public static ModelConfig loadModelConfig(String path, SourceType sourceType) throws IOException {
-        return loadJSON(path, sourceType, ModelConfig.class);
+        ModelConfig modelConfig = loadJSON(path, sourceType, ModelConfig.class);
+        if(modelConfig.getTrain().getGridConfigFile() != null) {
+            String gridConfigPath = modelConfig.getTrain().getGridConfigFile();
+            if(sourceType.equals(SourceType.HDFS)) {
```
Comment (zhangpengshan) [inline src/main/java/ml/shifu/shifu/util/CommonUtils.java]: sourceType == SourceType.HDFS is better

File: src/main/java/ml/shifu/shifu/util/CommonUtils.java [hunk 10]
```diff
@@ -248,7 +256,46 @@ public static ModelConfig loadModelConfig() throws IOException {
...
+                // gridsearch config file is uploaded to modelset path
+                gridConfigPath = new PathFinder(modelConfig).getPathBySourceType(
+                        gridConfigPath.substring(gridConfigPath.lastIndexOf(File.separator) + 1), SourceType.HDFS);
```
Comment (zhangpengshan) [inline src/main/java/ml/shifu/shifu/util/CommonUtils.java]: If grid configuration file doesn't have File.separator, exception here?

File: src/main/java/ml/shifu/shifu/util/CommonUtils.java [hunk 11]
```diff
@@ -325,7 +372,8 @@ private static void checkPathAndMode(String path, SourceType sourceType) {
      *             if any IO exception in parsing json.
      */
     public static List<ColumnConfig> loadColumnConfigList() throws IOException {
-        List<ColumnConfig> columnConfigList = loadColumnConfigList(Constants.LOCAL_COLUMN_CONFIG_JSON, SourceType.LOCAL);
+        List<ColumnConfig> columnConfigList = loadColumnConfigList(Constants.LOCAL_COLUMN_CONFIG_JSON,
+                SourceType.LOCAL);
```
Comment (zhangpengshan) [inline src/main/java/ml/shifu/shifu/util/CommonUtils.java]: looks your formater is not set well, by default, we use 120 chars per line

**Pull request merged**